File: src/cpp/multithreading/14_reader_writer_lock.cpp
#include <chrono>
#include <iostream>
#include <mutex>
#include <random>
#include <thread>
#include <vector>

class SharedData {
public:
  void read(int thread_id) {
    {
      std::unique_lock<std::mutex> lock(readers_mutex);
      ++readers;
      if (readers == 1) {
        data_mutex.lock();
      }
    }

    std::cout << "Reader " << thread_id << " reads data: " << data << std::endl;
    std::this_thread::sleep_for(
        std::chrono::milliseconds(random_duration(500, 1000)));

    {
      std::unique_lock<std::mutex> lock(readers_mutex);
      --readers;
      if (readers == 0) {
        data_mutex.unlock();
      }
    }
  }

  void write(int thread_id) {
    std::unique_lock<std::mutex> lock(data_mutex);
    std::cout << "Writer " << thread_id << " is writing..." << std::endl;
    ++data;
    std::this_thread::sleep_for(
        std::chrono::milliseconds(random_duration(500, 1000)));
    std::cout << "Writer " << thread_id
              << " finished writing. New data: " << data << std::endl;
  }

  int random_duration(int min, int max) {
    std::random_device rd;
    std::mt19937 mt(rd());
    std::uniform_int_distribution<int> dist(min, max);
    return dist(mt);
  }

private:
  int data = 0;
  std::mutex data_mutex;
  int readers = 0;
  std::mutex readers_mutex;
};

void reader(SharedData &shared_data, int thread_id) {
  while (true) {
    shared_data.read(thread_id);
    std::this_thread::sleep_for(
        std::chrono::milliseconds(shared_data.random_duration(1000, 2000)));
  }
}

void writer(SharedData &shared_data, int thread_id) {
  while (true) {
    shared_data.write(thread_id);
    std::this_thread::sleep_for(
        std::chrono::milliseconds(shared_data.random_duration(1000, 2000)));
  }
}

int main() {
  SharedData shared_data;

  const int num_readers = 3;
  const int num_writers = 2;

  std::vector<std::thread> threads;

  for (int i = 0; i < num_readers; ++i) {
    threads.emplace_back(reader, std::ref(shared_data), i);
  }

  for (int i = 0; i < num_writers; ++i) {
    threads.emplace_back(writer, std::ref(shared_data), i);
  }

  for (auto &thread : threads) {
    thread.join();
  }

  return 0;
}

File: src/python/multithreading/14_reader_writer_lock.py
"""
This script demonstrates the use of a Reader-Writer lock to manage concurrent access
to shared data in a multi-threaded environment. The Reader-Writer lock allows multiple
readers to access the shared data simultaneously, while preventing concurrent access
by readers and writers. This can be particularly useful when read operations are more
frequent than write operations, as it helps reduce contention between reader threads.

In this example, we implement a simple Reader-Writer lock using two standard locks: one
for managing the number of active readers and another for controlling access to the
shared data. We create separate reader and writer threads, which repeatedly perform
read and write operations on the shared data. The Reader-Writer lock ensures that
writers have exclusive access to the shared data when writing, and that readers can
read the data concurrently when there are no active writers.

Alternative solutions include using specialized Reader-Writer lock implementations
available in third-party libraries, or using other synchronization primitives such as
semaphores or condition variables.
"""

import threading
import time
import random


class SharedData:
    def __init__(self):
        self.data = 0
        self.lock = threading.RLock()
        self.readers = 0
        self.readers_lock = threading.Lock()

    def read(self, thread_id):
        with self.readers_lock:
            self.readers += 1
            if self.readers == 1:
                self.lock.acquire()

        print(f"Reader {thread_id} reads data: {self.data}")
        time.sleep(random.uniform(0.5, 1))  # Simulate some work

        with self.readers_lock:
            self.readers -= 1
            if self.readers == 0:
                self.lock.release()

    def write(self, thread_id):
        with self.lock:
            print(f"Writer {thread_id} is writing...")
            self.data += 1
            time.sleep(random.uniform(0.5, 1))  # Simulate some work
            print(f"Writer {thread_id} finished writing. New data: {self.data}")


def reader(shared_data, thread_id):
    while True:
        shared_data.read(thread_id)
        time.sleep(random.uniform(1, 2))


def writer(shared_data, thread_id):
    while True:
        shared_data.write(thread_id)
        time.sleep(random.uniform(1, 2))


def main():
    shared_data = SharedData()

    num_readers = 3
    num_writers = 2

    threads = []

    for i in range(num_readers):
        t = threading.Thread(target=reader, args=(shared_data, i))
        t.start()
        threads.append(t)

    for i in range(num_writers):
        t = threading.Thread(target=writer, args=(shared_data, i))
        t.start()
        threads.append(t)

    for t in threads:
        t.join()


if __name__ == "__main__":
    main()
